sim/team-validator.ts

```diff
@@ -9,6 +9,155 @@
 
 import {Dex} from './dex';
 
+/**
+ * Describes a possible way to get a pokemon. Is not exhaustive!
+ * sourcesBefore covers all sources that do not have exclusive
+ * moves (like catching wild pokemon).
+ *
+ * First character is a generation number, 1-7.
+ * Second character is a source ID, one of:
+ *
+ * - E = egg, 3rd char+ is the father in gen 2-5, empty in gen 6-7
+ *   because egg moves aren't restricted to fathers anymore
+ * - S = event, 3rd char+ is the index in .eventPokemon
+ * - D = Dream World, only 5D is valid
+ * - V = Virtual Console transfer, only 7V is valid
+ *
+ * Designed to match MoveSource where possible.
+ */
+export type PokemonSource = string;
+
+/**
+ * Represents a set of possible ways to get a Pokémon with a given
+ * set.
+ *
+ * `new PokemonSources()` creates an empty set;
+ * `new PokemonSources(dex.gen)` allows all Pokemon.
+ */
+export class PokemonSources {
+	/**
+	 * A list of specific possible PokemonSources
+	 */
+	sources: PokemonSource[];
+	/**
+	 * if nonzero: the set also contains all possible sources from
+	 * this gen and earlier.
+	 */
+	sourcesBefore: number;
+	/**
+	 * the set requires sources from this gen or later
+	 * this should be unchanged from the format's minimum past gen
+	 * (3 in modern games, 6 if pentagon is required, etc)
+	 */
+	sourcesAfter: number;
+	babyOnly?: string;
+	sketchMove?: string;
+	hm?: string;
+	restrictiveMoves?: string[];
+	/** Obscure learn methods */
+	restrictedMove?: ID;
+	/**
+	 * `limitedEgg` tracks moves that can only be obtained from an egg with
+	 * another father in gen 2-5. If there are multiple such moves,
+	 * potential fathers need to be checked to see if they can actually
+	 * learn the move combination in question.
+	 *
+	 * The custom value `'self'` is always incompatible.
```

Comment:
`'self'` is specifically for ExtremeSpeed Dragonite, for with the mother and father must both be Dragonite, and is thus incompatible with the father being anything else.

Hence, the father is itself: Dragonite.

Maybe I should just allow same-evo-family breeding, instead of using this weird workaround. I forgot why I did it in the first place.

Reply:
Oh, I remember, because compatibility issues; you always have everything in your moveset. "You can get Head Smash as a Relicanth egg move by breeding with a Relicanth with Head Smash" is not actually enough to prove that Head Smash is legal.

Reply:
I think my main confusion is that its a list of move names, and `self` doesn't feel like a 'move'? I mean, the whole point is that its some sentinel value, so I guess it doesn't need to be a move at all, but it just feels like a different 'kind' here.

Reply:
Yeah, it _is_ some sentinel value. Maybe it should just be `null`.